tensorflow/python/layers/base.py

```diff
@@ -450,9 +453,61 @@ def _should_add_regularizer(variable, existing_variable_set):
             **kwargs)
 
         if regularizer:
-          if (ops.executing_eagerly_outside_functions()
-              or _should_add_regularizer(variable, existing_variables)):
-            self._handle_weight_regularization(name, variable, regularizer)
+
+          # In order to support DistributionStrategy, we move regularization
+          # in the following `add_regularizer` merge_fn. Regularization should
+          # be declared according to variable type. For example, we should call
+          # regularizer on all repilcas when it is a `MirroredVariable`.
+          # However, `MirroredVariable` is created when first replica thread is
+          # called, which makes regularizer statements on other replica
+          # thread ignored because of the `existing_variabale` check. Therefore,
+          # it is reasonable to put regularization statement in
+          # `cross_replica` context.
```

Comment:
I am not convinced that this is the solution. Won't the value of the regularization loss be the same on all replicas because it is a function of the variable which is mirrored?

Reply:
Thanks for your comments. The problem is In current `MirroredStrategy`, regularization loss will not be declared with `MirroredVariable` at the same time. You should pay attention to the `existing_variable`, it contains `MirroredVariable` fetched from tf.global_variables. So, the first replica will create `MirroredVariable` and put them into tf.global_variables list. And the second replica will see all `MirroredVariable` in `existing_variables` because they have been put by first replica! Therefore, first replica does have regularization_loss, but other replicas does not.

It has performance issue in the following code in `Estimator`, the performance will be bad.
```
def model_fn():
    regularization_losses = tf.get_collection(
              tf.GraphKeys.REGULARIZATION_LOSSES)
    regularization_loss = tf.add_n(regularization_losses)
```
Of course the `model_fn` will be called on each replica to construct reg_loss and the result is right, but the regularization nodes are placed on replica 0, so other replicas will fetch them from replica 0 which involve unnecessary memcpy. 

The reason I put them in the `merge_call` is that I want to declare multiple `regularizer` on multiple devices. If I don't put them into `merge_call`, other replicas will loss the chance to declare their own `regularizer` because of the `MirroredVariable` in `existing_variables`.

I know the `global_collections` will be deprecated in TF2.0, but still some users are eager to use 1.0 api like tf.layers. That's the reason why I fix this bug in this way.

Reply:
Thanks for your comments @anj-s . The problem is In current MirroredStrategy, regularizer will not be declared with `MirroredVariable` declaration at the same time. You should pay attention to the existing_variable, it contains MirroredVariable fetched from tf.global_variables. So, the first replica will create MirroredVariable and put them into tf.global_variables list. And the second replica will see all MirroredVariable in existing_variables because they have been put by first replica! Therefore, first replica does have regularization_loss, but other replicas does not.

It has performance issue in the following code in Estimator, the performance will be bad.

def model_fn():
    regularization_losses = tf.get_collection(
              tf.GraphKeys.REGULARIZATION_LOSSES)
    regularization_loss = tf.add_n(regularization_losses)
Of course the model_fn will be called on each replica to construct reg_loss and the result is right, but the regularization nodes are placed on replica 0, so other replicas will fetch them from replica 0 which involve unnecessary memcpy.

The reason I put them in the merge_call is that I want to declare multiple regularizer on multiple devices. If I don't put them into merge_call, other replicas will loss the chance to declare their own regularizer because of the MirroredVariable in existing_variables.

I know the global_collections will be deprecated in TF2.0, but still some users are eager to use 1.0 api like tf.layers. That's the reason why I fix this bug in this way.